Close with error

erikh opened the pull request:
This patch augments the interfaces' close methods with an error return which is
only actually useful in one or two spots for now. However, it can be utilized
later as things fill in.

Fixes #82

mtrmac commented on docker/daemon/daemon_dest.go:
```diff
@@ -91,9 +91,11 @@ func imageLoadGoroutine(ctx context.Context, c *client.Client, reader *io.PipeRe
+	var err error
 	if !d.committed {
 		logrus.Debugf("docker-daemon: Closing tar stream to abort loading")
+		err = errors.New("Aborting upload, daemonImageDestination closed without a previous .Commit()")
```
This is not an error in `.Close()`; if the caller decides to abort without a `.Commit()`, and the upload is successfully aborted, `.Close()` should indicate success.

mtrmac replied:
This is still outstanding.

mtrmac commented on types/types.go:
```diff
@@ -139,7 +139,7 @@ type ImageDestination interface {
 	Reference() ImageReference
 	// Close removes resources associated with an initialized ImageDestination, if any.
-	Close()
+	Close() error
```
The caller in `copy.Image` should be updated to handle the error.

mtrmac commented on types/types.go:
```diff
@@ -185,7 +185,7 @@ type UnparsedImage interface {
 	Reference() ImageReference
 	// Close removes resources associated with an initialized UnparsedImage, if any.
-	Close()
+	Close() error
```
The callers in `copy.Image` (both for `UnparseImage` and `Image`) should be updated to handle the error.

And perhaps `doc.go` as well?

mtrmac commented:
To link the two, this fixes #82.

erikh commented:
Will make adjustments, and file a PR on skopeo in the next few days.

erikh commented:
Ok I updated this to handle close errors, I also changed the return convention (read the comments at the inner-top of copy.Image) to use output variables. All is explained in comments, but if this feels too cumbersome for the benefits we can talk about alternatives.

mtrmac commented on copy/copy.go:
```diff
@@ -94,8 +94,15 @@ type Options struct {
+// source image admissibility.
+func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageReference, options *Options) (retErr error) {
+	// XXX this function uses an output parameter for the error return value.
+	// Setting this and returning is the ideal way to return an error.
```
Looking at https://golang.org/ref/spec#Return_statements , even with a named return value it would be perfectly fine to just `return somerrorvalue` AFAICS.

(Also, doesn’t `XXX` mean “this is bad”? Perhaps just `NOTE:` or something like that.)

erikh replied:
Sure. I typically use it to say "WARNING" or something of the like: "pay attention" basically.

mtrmac commented on copy/copy.go:
```diff
@@ -106,9 +113,15 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
-	defer dest.Close()
+	defer func() {
+		if err := dest.Close(); err != nil {
+			retErr = errors.Wrapf(retErr, "dest: %v", err)
```
This should only happen if `retErr == nil`; the original value of `retErr` was the primary cause we are failing; that `dest.Close()` has failed may be completely unrelated, or it may be a side effect of the original `retErr` failure.

erikh replied:
ok sure, I was just thinking that maybe it'd be nice to have all the error context in the event of a copy failure.

mtrmac replied:
Oh, sorry, I wasn’t paying enough attention to notice that this preserves both. Let me check…

mtrmac replied:
… something like this would be nice _if we can make that human-readable_. See how the other `errors.Wrapf` calls in here try to tell the user what is going on. I am not sure that this is workable here:
```
errors.Wrapf(retErr, "Error closing destination: %v, after primary failure", err)
```
is bordering on incomprehensible, especially if `err` were long / created by several `errors.Wrapf` calls. (We can make the underlying data type situation better by using something like `aggregateErr` from `openshift-copies.go`, but that is probably even worse for readability.)

If we can’t make that human-readable, I think preserving a non-`nil` `retErr` is more important than the failure from `.Close`.

@runcom ?

erikh replied:
Sure, I get the concern.

What if I did something like this:

`errors.Wrapf(retErr, "(destination error %v) ", err)`

That way each indepednent error would be wrapped by parens. I could also add newlines.

I think the errors will be valuable context in the face of other, higher level errors.

mtrmac replied:
That’s certainly better than my attempt.

The code as is (making `daemonImageSource.Close` fail using an invalid path):
```
$ sudo ./x docker-daemon:busybox:latest dir:d
FATA[0000] Error copying: source: remove makethisfail//var/tmp/docker-daemon-tar266436759: no such file or directory: Error writing blob: open d/dir-put-blob135662858: no such file or directory 
```
quite misleading about the cause/effect relationships

“primary error” wording:
```
FATA[0000] Error copying: Error closing source: remove makethisfail//var/tmp/docker-daemon-tar088311212: no such file or directory, after primary failure: Error writing blob: open d/dir-put-blob812853531: no such file or directory 
```
Still focusing on the wrong thing (closing failure)

With the parentheses:
```
FATA[0000] Error copying: (source error remove makethisfail//var/tmp/docker-daemon-tar460112488: no such file or directory): Error writing blob: open d/dir-put-blob983238823: no such file or directory 
```
Certainly better. Good enough?

Newlines are attractive (they would even allow us to build a bulleted list :) ) but risky, the caller is fairly likely not expecting them.

At this point I am really not sure, I’ll let @runcom weigh in.

mtrmac commented on copy/copy.go:
```diff
@@ -118,7 +131,9 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
 		if unparsedImage != nil {
-			unparsedImage.Close()
+			if err := unparsedImage.Close(); err != nil {
+				retErr = errors.Wrapf(retErr, "unparsed: %v", err)
```
Same here

mtrmac commented on copy/copy.go:
```diff
@@ -131,10 +146,15 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
-	defer src.Close()
+	defer func() {
+		if err := src.Close(); err != nil {
+			retErr = errors.Wrapf(retErr, "source: %v", err)
```
Same here

cyphar commented on copy/copy.go:
```diff
@@ -106,9 +113,15 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
 	if err != nil {
-		return errors.Wrapf(err, "Error initializing destination %s", transports.ImageName(destRef))
+		retErr = errors.Wrapf(err, "Error initializing destination %s", transports.ImageName(destRef))
+		return
```
This change is not necessary. `return X` will set `retErr = x` implicitly.

erikh commented:
PTAL

changes:

* using parens for inner errors now
* retErr code was adjusted based on return semantics.

erikh commented:
https://github.com/containers/image/pull/240 is blocking this.

mtrmac commented:
LGTM pending tests. (Shouldn't `test-skopeo` work even after the API change? If not, the way we deal with API interdependencies is usually to

- Temporarily modify the corresponding skopeo PR to vendor from the proposed branch, show that _that_ passes tests
- Then merge the containers/image PR, ignoring the test failure
- Update the skopeo PR to vendor from the updated containers/image master again, and merge it.

mtrmac commented:
:+1:

erikh commented:
rebased

runcom commented:
@erikh sorry needs one last rebase before I merge this

erikh commented:
done

On Sat, Feb 25, 2017 at 3:31 AM, Antonio Murdaca <notifications@github.com>
wrote:

> @erikh <https://github.com/erikh> sorry needs one last rebase before I
> merge this
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/image/pull/236#issuecomment-282478214>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABJ6-qzkfc8JfWeurltFgnwDkS-3Iyyks5rgBElgaJpZM4L32DS>
> .
>

runcom commented:
:+1:
